### PR TITLE
feat(contract): link with GLPI item

### PR DESCRIPTION
**Pull request by Rom1-B**
### Changes description

Enable a contract to be linked to the ticket when it is created from the form.

image

### Checklist

Please check if your PR fulfills the following specifications:

- [ ] Tests for the changes have been added
- [ ] Docs have been added/updated

### References

!24332

Closes #N/A

### CODE REVIEW

File: inc/abstractitiltarget.class.php [hunk 1]
```diff
@@ -1263,6 +1277,73 @@ protected function showLocationSettings($rand) {
...
+      $users_questions = [];
+      foreach ($result as $question) {
+         if ($question['itemtype'] != 'Contract') {
+            continue;
+         }
+         $users_questions[$question['sname']][$question['id']] = $question['name'];
+      }
+      Dropdown::showFromArray('_contract_question', $users_questions, [
+         'value' => $this->fields['contract_question'],
+      ]);
```
Comment (btry) [inline inc/abstractitiltarget.class.php:1340]: There is a method to do such dropdown easier. Try this 

```suggestion
      PluginFormcreatorQuestion::dropdownForForm(
         $this->getForm(),
         [
            'fieldtype' => ['glpiselect'],
            'itemtype' => Contract::class
         ],
         'due_date_question',
         $this->fields['contract_question']
      );

```

File: inc/targetticket.class.php [hunk 2]
```diff
@@ -1069,6 +1089,86 @@ protected function setTargetLocation($data, $formanswer) {
...
+                     'ON' => [
+                        'answer' => 'plugin_formcreator_questions_id',
+                        'question' => 'id',
+                     ]
+                  ]
+               ],
+               'WHERE' => [
+                  'answer.plugin_formcreator_formanswers_id' => $form_answer_id,
+                  'question.fieldtype'                       => "glpiselect",
+               ],
```
Comment (btry) [inline inc/targetticket.class.php:1132]: itemtype in questions table helps you to improbe the SQL query. A 2nd snippet  below will become useless.
```suggestion
               'WHERE' => [
                  'answer.plugin_formcreator_formanswers_id' => $form_answer_id,
                  'question.fieldtype'                       => "glpiselect",
                  'question.itemtype'                       => Contract::class
               ],
```

File: inc/targetticket.class.php [hunk 3]
```diff
@@ -1069,6 +1089,86 @@ protected function setTargetLocation($data, $formanswer) {
...
+
+            foreach ($answers as $answer) {
+               // Decode dropdown settings
+               $question = PluginFormcreatorQuestion::getById($answer[PluginFormcreatorQuestion::getForeignKeyField()]);
+               $itemtype = $question->fields['itemtype'];
+
+               // Skip if not a dropdown on contracts
+               if ($itemtype !== Contract::class) {
+                  continue;
+               }
```
Comment (btry) [inline inc/targetticket.class.php:1147]: This becomes useless if the above suggestion is applied

File: tests/3-unit/PluginFormcreatorTargetTicket.php [hunk 4]
```diff
@@ -1713,4 +1726,289 @@ public function testSetTargetLocation($instance, $formanswer, $expected) {
 
       $this->integer((int) $output['locations_id'])->isEqualTo($expected);
    }
+
+   public function providerSetTargetContract_NotSet() {
```
Comment (btry) [inline tests/3-unit/PluginFormcreatorTargetTicket.php:1730]:
```suggestion
   public function providerSetTargetContract_notSet() {
```

File: tests/3-unit/PluginFormcreatorTargetTicket.php [hunk 5]
```diff
@@ -1713,4 +1726,289 @@ public function testSetTargetLocation($instance, $formanswer, $expected) {
...
+      return [
+         [
+            'instance'   => $instance1,
+            'formanswer' => $formAnswer1,
+            'expected'   => $item2->getID(),
+         ],
+      ];
+   }
+
+   public function providerSetTargetContract_nothing() {
```
Comment (btry) [inline tests/3-unit/PluginFormcreatorTargetTicket.php:1816]:
```suggestion
   public function providerSetTargetContract_Nothing() {
```

File: tests/3-unit/PluginFormcreatorTargetTicket.php [hunk 6]
```diff
@@ -1713,4 +1726,289 @@ public function testSetTargetLocation($instance, $formanswer, $expected) {
...
+      return [
+         [
+            'instance'   => $targetTicket,
+            'formanswer' => $formanswer,
+            'expected'   => 0
+         ],
+      ];
+   }
+
+   public function providerSetTargetContract_noTemplate() {
```
Comment (btry) [inline tests/3-unit/PluginFormcreatorTargetTicket.php:1841]:
```suggestion
   public function providerSetTargetContract_NoTemplate() {
```

File: tests/3-unit/PluginFormcreatorTargetTicket.php [hunk 7]
```diff
@@ -1713,4 +1726,289 @@ public function testSetTargetLocation($instance, $formanswer, $expected) {
...
+         ],
+      ];
+   }
+
+   public function providerSetTargetContract() {
+      return array_merge(
+         $this->providerSetTargetContract_nothing(),
+         $this->providerSetTargetContract_noTemplate(),
+         $this->providerSetTargetContract_FromTemplate(),
+      );
```
Comment (btry) [inline tests/3-unit/PluginFormcreatorTargetTicket.php:1993]: Some sub providers not added here.

**Review by btry — CHANGES_REQUESTED** [review]

**Review by AdrienClairembault — APPROVED** [review]

**Review by btry — APPROVED** [review]

**Pull request merged**